src/System.Security.Cryptography.Cng/src/Microsoft/Win32/SafeHandles/NCryptSafeHandles.cs

```diff
@@ -63,11 +63,38 @@ private enum OwnershipState
         /// </summary>
         private SafeNCryptHandle _holder;
 
+        private SafeHandle _parentHandle;
+
         protected SafeNCryptHandle() : base(IntPtr.Zero, true)
         {
             return;
         }
 
+        protected SafeNCryptHandle(IntPtr handle, SafeHandle parentHandle)
+            : base(IntPtr.Zero, true)
+        {
+            if (parentHandle == null)
+                throw new ArgumentNullException(nameof(parentHandle));
+            if (parentHandle.IsClosed || parentHandle.IsInvalid)
+                throw new ArgumentException(SR.Argument_Invalid_SafeHandleInvalidOrClosed, nameof(parentHandle));
+
+            bool success = false;
+            parentHandle.DangerousAddRef(ref success);
+            _parentHandle = parentHandle;
+
+            // Don't set the handle value until after parentHandle has been validated and persisted to a field,
+            // otherwise Dispose will try to call the underlying Free function.
+            SetHandle(handle);
+
+            // But if this handle value IsInvalid then we'll never call ReleaseHandle, which leaves the parent open
+            // forever.  Instead, release such a parent now.
+            if (IsInvalid)
+            {
+                _parentHandle.DangerousRelease();
+                _parentHandle = null;
+            }
```

Comment:
> since it's a public type someone could theoretically extend it and change the IsInvalid implementation to do something else, and then I'm helping them keep handles open permanently

Ah, good point!